src/Collections/Sequences/Actions.dfy

```diff
@@ -0,0 +1,261 @@
+*  
+*  Modifications and Extensions: Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "../../Wrappers.dfy"
+include "../../Action.dfy"
+include "Seq.dfy"
+
+module Actions {
+  import opened Wrappers
+  import opened A = Action
+  import opened Seq
+
+  method Map<A, R>(
+    action: Action<A, R>,
+    s: seq<A>
+  )
+    returns (res: seq<R>)
+    requires forall i | i in s :: action.Requires(i)
+    ensures |s| == |res|
+    ensures
+      forall i ::
+        && 0 <= i < |s|
+      ==>
+        action.Ensures(s[i], res[i])
+  {
+    var rs := [];
+    for i := 0 to |s|
+      invariant |s[..i]| == |rs|
+      invariant forall j ::
+        && 0 <= j < i
+      ==>
+        action.Ensures(s[j], rs[j])
+    {
+      var r := action.Invoke(s[i]);
+      rs := rs + [r];
+    }
+    return rs;
+  }
+
+  method MapWithResult<A, R, E>(
+    action: Action<A, Result<R, E>>,
+    s: seq<A>
+  )
+    returns (res: Result<seq<R>, E>)
+    requires forall i | i in s :: action.Requires(i)
+    ensures
+      res.Success?
+    ==>
+      |s| == |res.value|
+    ensures
+      res.Success?
+    ==>
+        (forall i ::
+          && 0 <= i < |s|
+        ==>
+          action.Ensures(s[i], Success(res.value[i])))
+  {
+    var rs := [];
+    for i := 0 to |s|
+      invariant |s[..i]| == |rs|
+      invariant forall j ::
+        && 0 <= j < i
+      ==>
+        action.Ensures(s[j], Success(rs[j]))
+    {
+      var maybeR := action.Invoke(s[i]);
+      if maybeR.Failure? {
+        return Failure(maybeR.error);
+      }
+      var r := maybeR.value;
+      rs := rs + [r];
+    }
+    return Success(rs);
+  }
+
+  method FlatMap<A, R>(
+    action: Action<A, seq<R>>,
+    s: seq<A>
+  )
+    // The ghost parts is returned to facilitate
+    // threading proof obligations.
+    // Idealy, it would be great to remove this
+    // and simply prove everything about `res`.
+    // However in practice this has proven to be difficult.
+    // Given how flexible FlatMap is,
+    // there may not be a prcatical general solution.
+    returns (res: seq<R>, ghost parts: seq<seq<R>>)
+    requires forall i | i in s :: action.Requires(i)
+    ensures
+      && |s| == |parts|
+      && res == Flatten(parts)
+      && (forall i :: 0 <= i < |s|
+      ==>
+        && action.Ensures(s[i], parts[i])
+        && multiset(parts[i]) <= multiset(res))
+  {
+    parts := [];
+    var rs := [];
+    for i := 0 to |s|
+      invariant |s[..i]| == |parts|
+      invariant forall j ::
+        && 0 <= j < i
+      ==>
+        && action.Ensures(s[j], parts[j])
+        && multiset(parts[j]) <= multiset(rs)
+      invariant Flatten(parts) == rs
+    {
+      var r := action.Invoke(s[i]);
+      rs := rs + r;
+      LemmaFlattenConcat(parts, [r]);
+      parts := parts + [r];
+    }
+    return rs, parts;
+  }
+
+  method FlatMapWithResult<A, R, E>(
```

Comment:
It's definitely bothering me that we need this and can't just use `Flatten(MapWithResult(action, s))` but I know you warned me it's not that simple. Let me play around with it a bit. :)

Reply:
You should. It may be that with #28 many things are easier.
The difficulty was crossing the flatten boundary.

I will also note, that there may be a runtime efficiency "thing" in here.
If one or the other of these is easier to fold into a lazy seq creation,
then we should go with that one :)